src/java/test/org/apache/zookeeper/test/CnxManagerTest.java

```diff
@@ -351,15 +351,55 @@ public void testSocketTimeout() throws Exception {
         LOG.info("Election port: " + port);
         Thread.sleep(1000);
 
-        Socket sock = new Socket();
-        sock.connect(peers.get(1L).electionAddr, 5000);
-        long begin = Time.currentElapsedTime();
-        // Read without sending data. Verify timeout.
-        cnxManager.receiveConnection(sock);
-        long end = Time.currentElapsedTime();
-        if((end - begin) > ((peer.getSyncLimit() * peer.getTickTime()) + 500)) Assert.fail("Waited more than necessary");
-        cnxManager.halt();
-        Assert.assertFalse(cnxManager.listener.isAlive());
+        try (Socket sock = new Socket()) {
+            sock.connect(peers.get(1L).electionAddr, 5000);
+            long begin = Time.currentElapsedTime();
+            // Read without sending data. Verify timeout.
+            cnxManager.receiveConnection(sock);
+            long end = Time.currentElapsedTime();
+            if ((end - begin) > ((peer.getSyncLimit() * peer.getTickTime()) + 500))
+                Assert.fail("Waited more than necessary");
+            cnxManager.halt();
+            Assert.assertFalse(cnxManager.listener.isAlive());
+        }
+    }
+
+    /*
+     * Test if a duplicate SID appears in the cluster
+     */
+    @Test
+    public void testSameSID() throws Exception {
+        QuorumPeer peer = new QuorumPeer(peers, peerTmpdir[2], peerTmpdir[2], peerClientPort[2], 3, 2, 2000, 2, 2);
+        QuorumCnxManager cnxManager = new QuorumCnxManager(peer);
+        QuorumCnxManager.Listener listener = cnxManager.listener;
+        if (listener != null) {
+            Thread.UncaughtExceptionHandler handler = new Thread.UncaughtExceptionHandler() {
+                @Override
+                public void uncaughtException(Thread th, Throwable ex) {
+                    if (ex instanceof RuntimeException) {
+                        String msg = ex.getMessage();
+                        LOG.error(msg);
+                        Assert.assertEquals("org.apache.zookeeper.server.quorum.QuorumPeerConfig$ConfigException:" +
+                                " Appearing duplicate SID: 2", msg);
+                    }
+                }
+            };
+            listener.setUncaughtExceptionHandler(handler);
+            listener.start();
+        } else {
+            LOG.error("Null listener when initializing cnx manager");
+        }
+        try (Socket sock = new Socket()) {
+            InetSocketAddress electionAddr = peers.get(peer.getId()).electionAddr;
+            LOG.info("Creating socket connection, host: {}, port: {}",
+                    electionAddr.getHostString(), electionAddr.getPort());
+            sock.connect(electionAddr, 30000);
```

Comment:
I want to fix the `Connection Refused`. I thought the environment `jenkins` is not stable...